fakenet/listeners/ProxyListener.py

```diff
@@ -0,0 +1,397 @@
+            self.logger.debug('Proxy connected to listener on port %s' 
+                    % self.port)
+
+            while True:
+                readable, writable, exceptional = select.select([self.sock], 
+                        [], [], .001)
+                if not self.remote_q.empty():
+                    data = self.remote_q.get()
+                    self.sock.sendto(data, ('localhost', self.listener_port))
+                if readable:
+                    data = self.sock.recv(BUF_SZ)
+                    if data:
+                        self.listener_q.put(data)
+                    else:
+                        self.sock.close()
+                        exit(1)
+        except Exception as e:
+            self.logger.debug('Listener socket exception %s' % e)
+
+class ThreadedTCPServer(SocketServer.ThreadingMixIn, SocketServer.TCPServer):
+    pass
+
+class ThreadedUDPServer(SocketServer.ThreadingMixIn, SocketServer.UDPServer):
+    pass
+
+def get_top_listener(config, data, listeners, diverter, orig_src_ip, 
+        orig_src_port, proto):
+    
+
+    top_listener = None
+    top_confidence = 0
+    dport = diverter.getOriginalDestPort(orig_src_ip, orig_src_port, proto)
+
+    for listener in listeners:
+  
+        try:
+            confidence = listener.taste(data, dport)
+            if confidence > top_confidence:
+                top_confidence = confidence
+                top_listener = listener
+        except:
+            # Exception occurs if taste() is not implemented for this listener
+            pass
+    
+    return top_listener
+
+class ThreadedTCPRequestHandler(SocketServer.BaseRequestHandler):
+
+    
+    def handle(self):
+
+        remote_sock = self.request
+        # queue for data received from the listener
+        listener_q = Queue.Queue()
+        # queue for data received from remote
+        remote_q = Queue.Queue()
+        data = None
+
+        ssl_remote_sock = None
+        ssl_config = { 
+                'certfile': 'listeners/ssl_utils/server.pem', 
+                'keyfile': 'listeners/ssl_utils/privkey.pem',
+                'ssl_version' : ssl.PROTOCOL_SSLv23 }
+
+        try:
+            data = remote_sock.recv(BUF_SZ, socket.MSG_PEEK)
+            self.server.logger.debug('Received data\n%s' % hexdump.hexdump(data, 
+                result='return'))
+        except Exception as e:
+            self.server.logger.info('recv() error: %s' % e.message)
+
+        if data:
+
+            if ssl_detector.looks_like_ssl(data):
+                self.server.logger.debug('SSL detected')
+                ssl_remote_sock = ssl.wrap_socket(
+                        remote_sock, 
+                        server_side=True, 
+                        do_handshake_on_connect=True,
+                        certfile=ssl_config['certfile'], 
+                        ssl_version=ssl_config['ssl_version'],
+                        keyfile=ssl_config['keyfile'] )
```

Comment:
There is something wrong with keyfile path:

Exception happened during processing of request from ('172.16.163.138', 49173)
Traceback (most recent call last):
  File "C:\Program Files\Python27\lib\SocketServer.py", line 596, in process_req
uest_thread
    self.finish_request(request, client_address)
  File "C:\Program Files\Python27\lib\SocketServer.py", line 331, in finish_requ
est
    self.RequestHandlerClass(request, client_address, self)
  File "C:\Program Files\Python27\lib\SocketServer.py", line 652, in __init__
    self.handle()
  File "c:\program files\python27\lib\site-packages\fakenet_ng-1.2-py2.7.egg\fak
enet\listeners\ProxyListener.py", line 253, in handle
    keyfile=ssl_config['keyfile'] )
  File "C:\Program Files\Python27\lib\ssl.py", line 943, in wrap_socket
    ciphers=ciphers)
  File "C:\Program Files\Python27\lib\ssl.py", line 554, in __init__
    self._context.load_cert_chain(certfile, keyfile)
IOError: [Errno 2] No such file or directory